src/diffusers/workflow_utils.py

```diff
@@ -0,0 +1,136 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Module for managing workflows."""
+import os
+from typing import Callable, Dict, List, Union
+
+import numpy as np
+import torch
+
+from .configuration_utils import ConfigMixin
+from .utils import PushToHubMixin, logging
+from .utils.constants import WORKFLOW_NAME
+
+
+logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
+
+_NON_CALL_ARGUMENTS = ["lora", "is_torch_tensor_present", "_name_or_path", "_class_name", "_diffusers_version"]
+_ALLOWED_PATTERNS = r"^[\w\s.,!?@#$%^&*()_+-=<>[\]{}|\\;:'\"/]*$"
+
+
+def populate_workflow_from_pipeline(
+    argument_names: List[str], call_arg_values: Dict, pipeline_name_or_path: str
+) -> Dict:
+    r"""Populates the call arguments and (optional) LoRA information in a dictionary.
+
+    Args:
+        argument_names (`List[str]`): List of function arguments.
+        call_arg_values (`Dict`):
+            Dictionary containing the arguments and their values from the current execution frame.
+        pipeline_name_or_path (`str`): Name or the local path to the pipeline that was used to generate the workflow.
+
+    Returns:
+        `Dict`: A dictionary containing the details of the pipeline call arguments and (optionally) LoRA checkpoint
+        details.
+    """
+    workflow = Workflow()
+
+    # Populate call arguments.
+    call_arguments = {
+        arg: call_arg_values[arg]
+        for arg in argument_names
+        if arg != "return_workflow"
+        and "image" not in arg
+        and not isinstance(call_arg_values[arg], (torch.Tensor, np.ndarray, Callable))
+    }
+
+    workflow.update(call_arguments)
+
+    generator = workflow.pop("generator")
+    if generator is not None:
+        workflow.update({"generator_seed": generator.initial_seed()})
+        workflow.update({"generator_device": generator.device})
+    else:
+        workflow.update({"generator_seed": None})
+        workflow.update({"generator_device": "cpu"})
+
+    workflow["_name_or_path"] = pipeline_name_or_path
+
+    return workflow
+
+
+class Workflow(dict, ConfigMixin, PushToHubMixin):
+    """Class sub-classing from native Python dict to have support for interacting with the Hub."""
+
+    config_name = None
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.config_name = WORKFLOW_NAME
+        self._internal_dict = {}
+
+    def __setitem__(self, __key, __value):
+        self._internal_dict[__key] = __value
+        return super().__setitem__(__key, __value)
+
+    def update(self, __m, **kwargs):
+        self._internal_dict.update(__m, **kwargs)
+        super().update(__m, **kwargs)
+
+    def pop(self, __key):
+        self._internal_dict.pop(__key)
+        super().pop(__key)
```

Comment:
Sorry if not appropriate but I ended up reviewing the PR in the end :)

Out of curiosity, what is the purpose of `_internal_dict`? IMHO, if you inherit from `dict` it is precisely to avoid having to define the `__setitem__`, `update`, `pop`,... methods yourself. Also here a lot of magic dict methods are not implemented (`__getitem__`, `__contains__`, `get`, etc...) which creates an `_internal_dict` that will eventually end up in a corrupted state.

If `_internal_dict` has a specific purpose, I would add a comment to let future devs know about it :)

Reply:
Thank you for your time, Lucain! Appreciate it.

Long story short, `_internal_dict` is how we maintain the configuration variables across the pipeline components. See here as an example:

https://github.com/huggingface/diffusers/blob/f912f39b50f87e50a9d99346f5c1b6e644653262/src/diffusers/configuration_utils.py#L112

Reply:
Oh ok, thanks for the explanation! So nothing to do with the fact that it's inherited from `dict` in the end. I scanned the PR looking for `_internal_dict` but did not realized it was already used in `diffusers`. All good then :+1: